### PR TITLE
feat: require verified address

### PR DESCRIPTION
**Pull request by dadrus**
## Related issue

@aeneasr: This is a promised PR for #1328.

## Proposed changes

As discussed in #1328. The only difference is the configuration. Instead of making the hook available in `before` property, the `after` is used as the information about the identity is only available there.

```.yaml
flows:
    login:
      ui_url: http://127.0.0.1:9090/identity/login
      lifespan: 10m
      after:
        hooks:
          - hook: require_verified_address
```

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by dadrus** [comment]
@aeneasr: In addition to the documentation update (which will follow), I would like to clarify one thing: 

An identity can have multiple verifiable addresses. The current implementation of the PR expects all of these to be verified. This might be an issue. One could change the behavior to make it expect at least one verified address, but this might be problematic as well. In our case, we would like to have the eMail address used for the current login attempt verified. So maybe kind of a configuration with useful defaults is required. So something like this:
```.yaml
  hook: require_verified_address
  config:
    verify: all # or alternatively 'one', or 'current' 
```
If the config is omitted, `one` might be a useful default.

If we go this way, how can I get the information about the email address used for the login? But there are also other related topics. In our case, we also support logins with user names. In this case, the expectation is, that at least `one` email address is verified. This however would not be possible to configure given the above configuration approach.

Any ideas from your side?

**Comment by Zageron** [comment]
What happens if someone's session expires immediately after changing their primary email, and they misspell it? Blocked from logging in because email is unverified and unable to verify because wrong email? Does the email change flow have a verification component in order to successfully apply?

**Comment by radekg** [comment]
Interesting question the one above. I’d expect the app to not allow changing an email address if there is no another confirmed address on the account.

**Comment by aeneasr** [comment]
Hello there, this is a great contribution! Sorry for the late review but heading off to holidays and had a ton of stuff on my plate :)

So generally I would suggest to avoid the one / all / current mechanism. If we add more authentication methods, this will get blurier. What for example about OpenID Connect logins that do not have an email address associated?

I think the "one" strategy is good enough. If the user has verified an address I think it is safe to assume that the user has passed that challenge. I'm not sure if having multiple addresses be verified would add an additional benefit - on the contrary! Users might be wondering "why do I need to verify several email addresses?" or "I verified my email address why can I not log in?".

From my side - the one strategy as the default would be absolutely good enough! :)

**Comment by radekg** [comment]
On the other hand, if I signed up using an email and a password, I should be able to recover using one of the social logins. I actually did that once with meetup.com. I created an account using facebook sso but I deleted my facebook account and was unable to sign in. I was able to recover the account by signing up with google directly. This linked the account to the old one and all was good.

The second thought I have, isn't signing up with social login an implied verification? If I signed up with Google SSO, Google just told you that the identity is genuine. If someone else was able to sign up using my Google account, well, that was a game over because my Google account was already compromised. So a registration via SSO linking implies a verified address, IMO.

Enjoy your holidays.

**Comment by dadrus** [comment]
@aeneasr: Yeah, simpler is better :) I'll update the code accordingly next week and will also add documentation.

Enjoy the long weekend!

**Comment by dadrus** [comment]
@radekg: Maybe I miss something, but how do you see this PR is related to recovery?

> The second thought I have, isn't signing up with social login an implied verification?

**Comment by radekg** [comment]
> @radekg: Maybe I miss something, but how do you see this PR is related to recovery?

Sorry, I think I’ve mixed stuff up with https://github.com/ory/kratos/pull/1323#issuecomment-846232227

**Comment by dadrus** [comment]
@aeneasr: Please take a look. There are some questions (as comments), which I would like to clarify, before finalizing the implementation and based on that updating the documentation.

File: selfservice/flow/login/hook.go
Comment (aeneasr) [inline selfservice/flow/login/hook.go:78]: So `Active` refers to the method the user used to sign in / up / verify / ...

So once you e.g. click "sign in with username / password" this would be set to "password". If you use OIDC this would be "oidc". It's basically a hint for the UI implementor.

Therefore, I don't think we need to set it here!

**Review by aeneasr — COMMENTED** [review]
This looks great! Only thing we need to add is some documentation to: https://www.ory.sh/kratos/docs/next/self-service/hooks/#after

It probably also makes sense to only allow this hook for password flows? For OIDC flows I don't think it makes sense?

File: selfservice/flow/login/hook.go
Comment (dadrus) [reply selfservice/flow/login/hook.go:78]: This is exactly the point. `Active` is never set. If it would, I would not implemented the workaround after the comment

**Comment by dadrus** [comment]
> It probably also makes sense to only allow this hook for password flows? For OIDC flows I don't think it makes sense?

Yes, I was thinking to adjust the schema for this purpose. Will come with the next commits

File: selfservice/flow/login/hook.go
Comment (dadrus) [reply selfservice/flow/login/hook.go:78]: @aeneasr: Since `Active` is never set, I consider it as a bug and will fix it as part of the next commits.

**Comment by dadrus** [comment]
@aeneasr: Please take a look. If I haven't forgotten anything, this PR should be complete.

**Review by aeneasr — APPROVED** [review]

**Comment by dadrus** [comment]
@aeneasr: I need your help. This is first time I have to deal with JS and cypress. I tried to implement the test cases, but unfortunately there were issues running the registration scenario locally. Only after completely rewriting the `register` command in `commands.js` the registration was successful. This way I was able to finalize the negative test. But now I run into issues while trying to verify the email in a successful case. The mailslurper just doesn't receive any emails and I don't understand why. Could you please support here? Thank you in advance.

File: test/e2e/cypress/support/commands.js
Comment (aeneasr) [inline test/e2e/cypress/support/commands.js:106]: This looks good! However, I think this might need the negative as well? So iterating over `login.['after'].password.hooks` and `filter` out any hooks that are `require_verified_address`

File: test/e2e/cypress/support/commands.js
Comment (aeneasr) [inline test/e2e/cypress/support/commands.js:103]: Instead of a replacement this should probably be adding the hook to the array, not replacing it :)

File: test/e2e/cypress/support/commands.js
Comment (aeneasr) [inline test/e2e/cypress/support/commands.js:97]: Similar to my comment below this should probably have a filter. Also, we want the inverse (so add session back in)

File: test/e2e/cypress/integration/profiles/verification/login/success.spec.js
Comment (aeneasr) [inline test/e2e/cypress/integration/profiles/verification/login/success.spec.js:23]: These tests need a `afterEach()` or `after()` which resets `cy.dontLoginUserAfterRegistration` and `cy.enableLoginForVerifiedAddressOnly` to the previous values or all subsequent tests will use this set up and thus fail.

File: test/e2e/cypress/integration/profiles/verification/login/errors.spec.js
Comment (aeneasr) [inline test/e2e/cypress/integration/profiles/verification/login/errors.spec.js:23]: See comment in `success.spec.js`

File: test/e2e/cypress/support/commands.js
Comment (aeneasr) [inline test/e2e/cypress/support/commands.js:135]: Changing the logic here will break all other tests. Please revert any changes to this method! I don't believe you need any of those here.

**Comment by dadrus** [comment]
The tests are not final, as there were two issues, I've written in https://github.com/ory/kratos/pull/1355#issuecomment-858553414. I appreciate, if you could help with those

**Comment by Quasarman** [comment]
Hello,

First of all thanks for the efforts for implementing this behavior!
Are there any further updates when this will become available?

BR

**Comment by aeneasr** [comment]
While the PR is being worked on I will mark it as a draft. That declutters our review backlog :)

Once you're done with your changes and would like someone to review them, mark the PR as ready and request a review from one of the maintainers.

Thank you!

**Comment by dadrus** [comment]
@Quasarman: I was on vacation until yesterday and before it, there were some urgent topics in a project I'm working for, which I had to address.
The actual implementation is actually complete. The things which are missing are the e2e tests, which I will address asap

**Comment by Quasarman** [comment]
Hello @dadrus ,

no worries thanks a lot for your efforts!

**Comment by aeneasr** [comment]
Nice, I hope you enjoyed your days off @dadrus :)

**Comment by dadrus** [comment]
Definitely :) Recharging batteries feels always good ;)

**Comment by dadrus** [comment]
@aeneasr: There are no much changes in the last commit - only some simplifications compared to what you've done during your call.

**Comment by aeneasr** [comment]
@dadrus yes you are right, looks like we don't need the reset method at all :)

**Comment by aeneasr** [comment]
Sorry for the breaking test - I had a fix locally but looks like you were faster 😅

**Comment by dadrus** [comment]
@aeneasr: If you're not going to break any further tests in a review :), I would say I've done everything required for this feature. Last commits have a self explanatory commit comments :), so feel free to merge or ask for further updates ;)

**Comment by aeneasr** [comment]
yeah sorry about that again 😓 also LGTM from my side

**Comment by dadrus** [comment]
Don't worry ;)

**Comment by aeneasr** [comment]
🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳

**Pull request merged**
